app/code/Magento/InventoryElasticsearch/Plugin/CatalogSearch/Model/Indexer/Fulltext/Action/DataProvider/StockedProductFilterByInventoryStock.php

```diff
@@ -0,0 +1,143 @@
+     *
+     * @param DataProvider $dataProvider
+     * @param array $indexData
+     * @param array $productData
+     * @param int $storeId
+     * @return array
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function beforePrepareProductIndex(
+        DataProvider $dataProvider,
+        array $indexData,
+        array $productData,
+        int $storeId
+    ): array {
+        if ($this->config->isElasticsearchEnabled() && !$this->stockConfiguration->isShowOutOfStock($storeId)) {
+            $productIds = array_keys($indexData);
+            $stockStatuses = [];
+
+            foreach ($this->getProductIdsByWebsiteIds($productIds) as $websiteId => $productIds) {
+                $stock = $this->stockByWebsiteIdResolver->execute($websiteId);
+                $stockTable = $this->stockIndexTableNameResolver->execute((int)$stock->getStockId());
+
+                if ($this->resourceConnection->getConnection()->isTableExists($stockTable)) {
+                    $connection = $this->resourceConnection->getConnection();
+                    $select = $connection->select();
+                    $select->from(
+                        ['product' => $this->resourceConnection->getTableName('catalog_product_entity')],
+                        ['entity_id']
+                    );
+                    $select->joinInner(
+                        ['stock' => $stockTable],
+                        "product.sku = stock.sku",
+                        ['is_salable']
+                    );
+                    $select->where('product.entity_id IN (?)', $productIds);
+                    $productsSalable = $connection->fetchAssoc($select);
+                    $productsSalable = array_filter($productsSalable, function ($productSalable) {
+                        return StockStatusInterface::STATUS_IN_STOCK == $productSalable['is_salable'];
+                    });
+                    $stockStatuses += $productsSalable;
+                }
+            }
+
+            $indexData = array_intersect_key($indexData, $stockStatuses);
+        }
+
+        return [
+            $indexData,
+            $productData,
+            $storeId,
+        ];
+    }
+
+    /**
+     * Get all website ids by product ids.
+     *
+     * @param array $entityIds
+     * @return array
+     */
+    private function getProductIdsByWebsiteIds(array $entityIds): array
+    {
+        $result = [];
+
+        $connection = $this->resourceConnection->getConnection('indexer');
+        $select = $connection->select();
+        $select->from(
+            ['product_in_websites' => $this->resourceConnection->getTableName('catalog_product_website')],
+            ['website_id', 'product_id']
+        )->where('product_in_websites.product_id IN (?)', $entityIds)->distinct();
+        foreach ($connection->fetchAll($select) as $websiteData) {
+            $result[(int)$websiteData['website_id']][] = (int)$websiteData['product_id'];
```

Comment:
probably it's more performance efficient to use GROUP  BY + GROUP CONCAT on DB level